### ISSUE TITLE
Could not move temporary workspace ... to immutable location

### ISSUE THREAD

**Issue by CC007** [gradle/gradle#31438]
### Current Behavior

Reopening #27844, because it seems to not be fixed and has been broken since 8.5

When trying to build a freshly created project, I get the following error:

```
FAILURE: Build failed with an exception.

* What went wrong:
A build operation failed.
    Could not move temporary workspace (F:\local\applicationData\gradle\caches\8.9\transforms\e7fa29f876e87b2a15c2bc90d0781fdd-9307ed84-c3b0-4017-bce5-8de12a7efb13) to immutable location (F:\local\applicationData\gradle\caches\8.9\transforms\e7fa29f876e87b2a15c2bc90d0781fdd)
> Could not move temporary workspace (F:\local\applicationData\gradle\caches\8.9\transforms\e7fa29f876e87b2a15c2bc90d0781fdd-9307ed84-c3b0-4017-bce5-8de12a7efb13) to immutable location (F:\local\applicationData\gradle\caches\8.9\transforms\e7fa29f876e87b2a15c2bc90d0781fdd)
```

This was even after already having moved GRADLE_USER_HOME off of my C drive (which some suggested could fix the issue).
The issue only occurs in version 8.6 and newer. Version 8.5 doesn't have this issue.

Manually I actually am able to do the move, but only after killing the Gradle daemon process, since it's locking the files.

**Workarounds:**

1. Some people mention that uninstalling the virus scanner might help. This however is obviously not a preferred solution, since we don't want Gradle users to be vulnerable to viruses.
2. As mentioned, you can manually move/rename gradle cashes as the error is trying to do (So `e7fa29f876e87b2a15c2bc90d0781fdd-9307ed84-c3b0-4017-bce5-8de12a7efb13` to `e7fa29f876e87b2a15c2bc90d0781fdd`. This works as a workaround, but is tedious (you need to do this at least 5-6 times before it works) and is only a temporary solution. Any Gradle version updates or cache wipes would require you to do this again.
3. Downgrade to version 8.5. This doesn't work for kotlin compose multiplatform projects though, since they rely on features from newer versions.

### Expected Behavior

I expect a simple new gradle project to just build without issue.

### Context (optional)

This behavior is not tied to a specific project. It was tested with a simple kotlin jvm application and a project generated by the compose multiplatform wizard.

### Self-contained Reproducer Project

The problem is not caused by the project itself (considering it happens with any Gradle project).
Here is the gradle project that I used in the build scan: https://github.com/CC007/test-gradle

### Gradle version

any version between 8.6 and 8.11.1

### Build scan URL (optional)

https://scans.gradle.com/s/n3yylnhnckwuy

### Your Environment (optional)

OS: Windows 10 (22H2 build 19045.5131)
Virus scanner: Bitdefender

**Comment by ov7a** [comment]
This issue needs a decision from the team responsible for that area. They have been informed. Response time may vary.

***

Related:
* #31392

**Comment by cakeGit** [comment]
Ive been having a similar issue, i noticed that you are also using bitdefender, could these be linked?

**Comment by CC007** [comment]
> Ive been having a similar issue, i noticed that you are also using bitdefender, could these be linked?

It might be related to Antivirus activity, but as said: it didn't happen in Gradle version 8.5. Something got changed in a way that triggers the write permission issues.

**Comment by pnemonic78** [comment]
> Ive been having a similar issue, i noticed that you are also using bitdefender, could these be linked?

My ".gradle" folder is already in bitdefender exceptions since ages ago.
Gradle 8.5 is working fine. Problem started with 8.6.
Manually rename the folder, e.g. "ren 81991bfd1b4ed8debe429166c1ad789e-6b703885-e467-4284-bbc7-e01c819eb8cd 81991bfd1b4ed8debe429166c1ad789e" helps.

**Comment by Azbesciak** [comment]
Hello, are there any plans to fix the problem? I struggle with that for 2 months at least

**Comment by sinuraya** [comment]
I just found out the easiest way to workaround this problem. All you have to do is delete the .\gradle folder altogether and let Android Studio rebuilt it for you.

**Comment by lptr** [comment]
The root cause in some cases seems to be:

```text
Caused by: java.nio.file.AccessDeniedException: C:\Users\Martin\.gradle\caches\transforms-4\07ecae6095036d727e29035ee34f7b31-9b217975-ba0b-4653-a335-76e7cf54574a -> C:\Users\Martin\.gradle\caches\transforms-4\07ecae6095036d727e29035ee34f7b31
	at org.gradle.internal.vfs.impl.DefaultFileSystemAccess.lambda$moveAtomically$12(DefaultFileSystemAccess.java:224)
	at org.gradle.internal.vfs.impl.DefaultFileSystemAccess.write(DefaultFileSystemAccess.java:212)
	at org.gradle.internal.vfs.impl.DefaultFileSystemAccess.moveAtomically(DefaultFileSystemAccess.java:223)
	at org.gradle.internal.execution.steps.AssignImmutableWorkspaceStep$WorkspaceMoveHandler.executeMoveOr(AssignImmutableWorkspaceStep.java:273)
	... 248 more
```

From https://github.com/gradle/gradle/issues/31542.

**Comment by asodja** [comment]
We have problems reproduce this issue, since it seems it happens under certain conditions and thus we have problem understanding the issue and implementing a good fix.

If anyone can reproduce it on their machine reliable. maybe by deleting `C:\path-to-gradle-cache\.gradle\caches\transforms` folder or some other way,  please contact us here or via [Gradle Community Slack](https://gradle.org/slack-invite/).  
We would like to get more logs via a custom distribution.

**Comment by Jimvy** [comment]
Hello! I have a computer which is regularly having this issue, and I'm open to try to help fixing this problem.

By "regularly", I mean that every 1 to 2 weeks, one or two Gradle projects, in a larger project containing 20 different Gradle projects, at random, fail to build with this particular error. This is still a bit random, but happens often enough.
With other colleagues, we've noticed that
- re-running the build multiple times may eventually fix the problem. Sometimes thrice, sometimes 20 times.
- deleting the folders doesn't help more than the random re-run.
- renaming the folders (when there is only one of them) sometimes helps the problem, according to a colleague.
- This doesn't happen on 8.5

Currently running with Gradle 8.9, I will try to use 8.12 in the upcoming days and see if this changes anything (good or bad).
This is a corporate laptop, Windows 10, with a bunch of antivirus/scanners included (no-one of which is BitDefender), without admin rights though, and unlikely to allow any executable to run (Jars should be fine), so there are limitations.

Happy to help! Cheers

**Comment by alejandroVieira19** [comment]
I solved this problem after deleting RAV endpoint.

I hope this help solving the problem for you guys.

**Comment by chubbard** [comment]
Here is what I'm seeing from running a single Spock functionalTest that used an `@TempDir` annotation in it on Gradle 8.10.2.  This worked in prior versions of Gradle of course.  If you want a reproducible example you can use this project's branch that can reproduce the problem:

https://github.com/chubbard/ecr-token-plugin/tree/release-1.0.2

The `release-1.0.2` branch is set to reproduce the problem.  You'll even get a detailed stacktrace showing the issue.

```
* What went wrong:
Error resolving plugin [id: 'com.fuseanalytics.gradle.ecrToken']
> A build operation failed.
      Could not move temporary workspace (D:\Users\charlie\projects\fuseanalytics\ecr-token-plugin\plugin\build\tmp\functionalTest\work\.gradle-test-kit\caches\8.10.2\transforms\c4b39ab3dec271b176aac6c953afa6a0-37b9547d-1eae-4d04-bd86-5e77d4d72a89) to immutable location (D:\Users\charlie\projects\fuseanalytics\ecr-token-plugin\plugin\build\tmp\functionalTest\work\.gradle-test-kit\caches\8.10.2\transforms\c4b39ab3dec271b176aac6c953afa6a0)
   > Could not move temporary workspace (D:\Users\charlie\projects\fuseanalytics\ecr-token-plugin\plugin\build\tmp\functionalTest\work\.gradle-test-kit\caches\8.10.2\transforms\c4b39ab3dec271b176aac6c953afa6a0-37b9547d-1eae-4d04-bd86-5e77d4d72a89) to immutable location (D:\Users\charlie\projects\fuseanalytics\ecr-token-plugin\plugin\build\tmp\functionalTest\work\.gradle-test-kit\caches\8.10.2\transforms\c4b39ab3dec271b176aac6c953afa6a0)
```
Now on the disk I have the folders:

1. c4b39ab3dec271b176aac6c953afa6a0-37b9547d-1eae-4d04-bd86-5e77d4d72a89
2. c4b39ab3dec271b176aac6c953afa6a0-8da1f971-9053-423b-af42-35fbe04dc940

You'll notice they start with the same number, and this is also the number you see in the error message as the destination folder (ie c4b39ab3dec271b176aac6c953afa6a0).  These 2 folders are the only folders with the suffix to distinguish them from each other.  Otherwise there would be a conflict trying having 2 folders with the same name.  

So here is my thought.  Some where in the code it is trying to create the same folder again, and it has to append this suffix.  And it's doing this twice thus the folder `c4b39ab3dec271b176aac6c953afa6a0-37b9547d-1eae-4d04-bd86-5e77d4d72a89` is maybe being renamed to `c4b39ab3dec271b176aac6c953afa6a0`  However, there already exists a folder that starts with `c4b39ab3dec271b176aac6c953afa6a0` so it calculates a suffix to distinguish, but it just happens to the be the same as `c4b39ab3dec271b176aac6c953afa6a0-8da1f971-9053-423b-af42-35fbe04dc940` which already exists and blammo it can't rename it.  However, the error message is misleading because it says `c4b39ab3dec271b176aac6c953afa6a0` as the destination folder, but really it's `c4b39ab3dec271b176aac6c953afa6a0-8da1f971-9053-423b-af42-35fbe04dc940`.  From the stacktrace I saw though this might not be accurate:

```
Caused by: java.nio.file.AccessDeniedException: D:\Users\charlie\projects\fuseanalytics\ecr-token-plugin\plugin\build\tmp\functionalTest\work\.gradle-test-kit\caches\8.10.2\transforms\8e882a9e27042921522a4f6a6be81ef8-70313f3e-833b-4416-8ed0-466ec56327b9 -> D:\Users\charlie\projects\fuseanalytics\ecr-token-plugin\plugin\build\tmp\functionalTest\work\.gradle-test-kit\caches\8.10.2\transforms\8e882a9e27042921522a4f6a6be81ef8
	at org.gradle.internal.vfs.impl.DefaultFileSystemAccess.lambda$moveAtomically$13(DefaultFileSystemAccess.java:229)
	at org.gradle.internal.vfs.impl.DefaultFileSystemAccess.write(DefaultFileSystemAccess.java:217)
	at org.gradle.internal.vfs.impl.DefaultFileSystemAccess.moveAtomically(DefaultFileSystemAccess.java:228)
	at org.gradle.internal.execution.steps.AssignImmutableWorkspaceStep$WorkspaceMoveHandler.executeMoveOr(AssignImmutableWorkspaceStep.java:273)
```

I have no actual proof of this hypothesis.  I haven't looked in the code.  But what sticks out to me is the 2 folders out of 87 that share the same prefix.  And all of these 87 folders showed up after I deleted by transforms directory in Gradle 8.10.2.

Hopefully now that I have a reproducible example the Gradle team can isolate the problem and fix it.

**Comment by chubbard** [comment]
Oh wait here is what transforms looks after the last time it failed.  This time I see lots of directories with shared prefixes and differing suffixes.  However, 1 pair where the directories that have this problem where the destination folder DOES exist which would cause the conflict!  See attached image.

![Image](https://github.com/user-attachments/assets/66202bb0-6ac6-4f71-b228-641feef1e792)

**Comment by lptr** [comment]
Thanks for trying to reproduce this, I'll take a closer look later this week.

> So here is my thought. Some where in the code it is trying to create the same folder again, and it has to append this suffix. And it's doing this twice thus the folder `c4b39ab3dec271b176aac6c953afa6a0-37b9547d-1eae-4d04-bd86-5e77d4d72a89` is maybe being renamed to `c4b39ab3dec271b176aac6c953afa6a0` However, there already exists a folder that starts with `c4b39ab3dec271b176aac6c953afa6a0` so it calculates a suffix to distinguish, but it just happens to the be the same as `c4b39ab3dec271b176aac6c953afa6a0-8da1f971-9053-423b-af42-35fbe04dc940` which already exists and blammo it can't rename it. However, the error message is misleading because it says `c4b39ab3dec271b176aac6c953afa6a0` as the destination folder, but really it's `c4b39ab3dec271b176aac6c953afa6a0-8da1f971-9053-423b-af42-35fbe04dc940`. From the stacktrace I saw though this might not be accurate:

Here's how the process works: when Gradle executes the transform, it generates content in a temporary directory with a long name like `c4b39ab3dec271b176aac6c953afa6a0-8da1f971-9053-423b-af42-35fbe04dc940`. Once the work is done Gradle tries to move the temporary directory to its final location in the cache with a short name (`c4b39ab3dec271b176aac6c953afa6a0`). This move is atomic, and ensures that any other daemons running at the same time cannot access a partial output.

On Windows, because of how the file system works, this move often fails. This typically happens when the original native process that created the output (like `node`) keeps a lock open on some files in the temporary directory. Another reason can be a virus scanner like Windows Defender looking at the content of the temporary directory while we are trying to move it.

In such a case Gradle copies the temporary directory to another temporary directory that the original process has no knowledge of (in this case `c4b39ab3dec271b176aac6c953afa6a0-37b9547d-1eae-4d04-bd86-5e77d4d72a89`), and attempts to move _that_ temporary directory to `c4b39ab3dec271b176aac6c953afa6a0` again. My theory is that a virus scanner or some other process manages to grab hold of some file in this new temporary directory, and prevents the move again. Gradle does not currently try to make the move a third time if this happens and instead produces the error message you see.

I see the following ways to address this:

1) We could try to work against virus scanners more. We could wait some time between retries, or retry the move multiple times. That won't be a complete fix, though, and the chance for a failed move will always be present.

2) The other option is to change how we store transform outputs on Windows, and instead of relying on atomic moves of directories, we would rely on file locks. The upside is that the process would be reliable. The downsides are added complexity and noticeably slower execution of work on Windows.

3) I don't think this is feasible, but we could also mandate that Gradle's user home directory always be exempt from virus scanning. We could check with popular scanners explicitly to see if they have the exemption configured. Again, I'm mentioning this only for completeness' sake, as I don't see it being feasible.

I'm currently leaning towards 2).

Another complicating factor is that we've seen some reports on Linux about similar-looking problems, but I think those have been addressed in Gradle 8.7 (IIRC). If anyone still sees a similar problem with Linux or macOS using Gradle 8.10 or later, please let us know, because then it's a different issue and we need to address that separately.

**Comment by chubbard** [comment]
So this plugin doesn't do anything with temporary directories.  That explanation makes sense for the functionalTest because of the temp directory usage.  However, all this plugin does is execute a call against AWS.  It doesn't create any files or anything like that, and I encountered this problem in my build when adding my plugin to a gradle 8.10.2 build.  If I don't have this plugin added it's fine.  I don't see the issue.  So how does that square with moving directories atomically?

**Comment by chubbard** [comment]
How did this behavior change in 8.6 that created these issues?  Prior version 8.5 didn't seem to suffer from this problem.  But executing tasks, probably with temp directories for output, has been apart of Gradle since like...forever right?

**Comment by asodja** [comment]
Before Gradle 8.6 we used "file lock" approach. But due to some extra work we added in 8.6, we had some regression with integration tests where a lot of daemons were running in the same Gradle user home (which is not really normal use case). 

We decided to implement a new approach that has better performance, but on Windows there are other problems with it. The issue is actually unrelated to tasks, but it's normally caused by artifact transforms that cache result in Gradle user home.

**Comment by stan-at-work** [comment]
Is there any workaround for this ATM ? I'm facing this when building for flutter.

**Comment by chubbard** [comment]
Maybe.  I've reverted to 8.5 for the time being until this is fixed.  It is intermittent but it happens enough that I can't be productive and run the latest gradle.  It only affects Windows users usually who have active virus protection on their systems.  The active scanning and poor Windows support for the changes Gradle team made in 8.6 trigger this defect.

So workarounds can include: 

1. Running on non-Windows platforms
2. Turning off active Virus scanning when you change files on your HD
3. Downgrade to Gradle 8.5

I wouldn't want to say turn off anti-virus since that could leave you exposed to security issues, and not all environment are able to control their virus scanners.

**Comment by jullicer** [comment]
I recently encountered an issue that took a while to track down. I tried several things, including:

- Downgrading Gradle to 8.2  
- Disabling Bitdefender  
- Modifying `gradle.properties`  
- Changing the Gradle cache directory  

None of these worked.

Eventually, I discovered the problem was caused by the Flutter and Dart extensions in my IDE. Here's what finally resolved it for me:

1. Downloaded a fresh copy of Flutter to `C:\flutter`  
2. Uninstalled the Flutter and Dart extensions from VS Code  
3. Relaunched the project — and voilà, it worked!

Hopefully, this helps someone else facing a similar issue.

**Comment by mauseoluwasegun** [comment]
please im facing the issue over 1 month now:



![Image](https://github.com/user-attachments/assets/bc8361e1-5c1d-4c4d-a6d1-233a2f17bce6)

**Comment by stan-at-work** [comment]
Any updates ? no build can be made on new generated flutter projects....

**Comment by gabriel7419** [comment]
Any updates? I'm also on Bitdefender. Can't disable it because I'm on the company's computer. Need help, please.

**Comment by HunterLomo** [comment]
This problem has been bothering me for a week. After trying various methods, I found that only the **8.5** version works normally.

**Comment by stan-at-work** [comment]
> This problem has been bothering me for a week. After trying various methods, I found that only the **8.5** version works normally.

Most likey a file lock constraint or anti virus lock. In my case btw

**Comment by mlopatkin** [comment]
The issue is in the backlog of the relevant team and is prioritized by them.

***

- We should handle this sooner
- Probably we should try to implement some windows-specific mode that doesn't rely on atomic renames.

**Comment by felipematheusmello** [comment]
For me it happened because of the Windows path pattern, which is not the same as in the git bash I often use (it gets the path like in a linux machine). Keeping it in mind, try to use PowerShell and run a ./gradlew clean to make sure that works. Hope that helps anyone that needs

**Comment by devfemibadmus** [comment]
If you have tried everything and you feel like giving up please try this
```python

import os
import shutil

base = r"C:\Users\Femi.Badmus\.gradle\caches\8.14\transforms" #change to ur gradle path

for name in os.listdir(base):
    if "-" in name:
        src = os.path.join(base, name)
        dst = os.path.join(base, name.split("-", 1)[0])
        if os.path.isdir(src):
            if os.path.exists(dst):
                shutil.rmtree(dst)
                print(f"Deleted: {dst}")
            print(f"Renaming: {src} -> {dst}")
            os.rename(src, dst)

```

<img width="1238" height="664" alt="Image" src="https://github.com/user-attachments/assets/8dc6677e-137e-433d-9611-15ef07121144" />

**Comment by chetan-kotwal** [comment]
I am also facing the same issue. Please suggest is there any alternative which can be used.

**Comment by ov7a** [comment]
@chetan-kotwal please open a new issue with a reproducer if you have the same issue.
